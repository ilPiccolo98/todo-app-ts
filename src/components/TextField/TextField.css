.primary-container {
  width: 150px;
  height: 120px;
  background-color: white;
  padding: 10px;
  border-radius: 10px;
}

.primary-label {
  font-size: 15px;
  color: black;
  font-family: sans-serif;
}

.secondary-container {
  width: 150px;
  height: 120px;
  background-color: gray;
  padding: 10px;
  border-radius: 10px;
}

.secondary-label {
  font-size: 15px;
  color: white;
  font-family: sans-serif;
}

.textField {
  border-radius: 5px;
}

.error {
  color: red;
  font-family: sans-serif;
  font-size: 15px;
}
