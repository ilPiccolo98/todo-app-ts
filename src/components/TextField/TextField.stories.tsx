import React from "react";
import { Meta, Story } from "@storybook/react";
import { TextFieldProps, TextField } from "./TextField";

const meta: Meta = {
  title: "TextField",
  component: TextField,
  argTypes: {
    labelText: {
      defaultValue: "Default label text",
    },
    placeholder: {
      defaultValue: "Default placeholder",
    },
    type: {
      defaultValue: "text",
    },
  },
};

export default meta;

const Template: Story<TextFieldProps> = (args) => <TextField {...args} />;

export const primary = Template.bind({});
primary.args = {
  variant: "primary",
  labelText: "primary",
  placeholder: "primary",
};

export const secondary = Template.bind({});
secondary.args = {
  variant: "secondary",
  labelText: "secondary",
  placeholder: "secondary",
};
