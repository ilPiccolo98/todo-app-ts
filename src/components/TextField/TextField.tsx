import React, { HTMLAttributes, KeyboardEvent, useState } from "react";
import "./TextField.css";
import MessageError from "../MessageError/MessageError";

export interface TextFieldProps extends HTMLAttributes<HTMLInputElement> {
  /**Which style the TextField component should have */
  variant: "primary" | "secondary";
  /**Text of the label*/
  labelText?: string;
  /**It determins the functianality of the text field */
  type: "text" | "password" | "numbers";
  /**Name of the TextField */
  name?: string;
  /**Value that should be content in the Textfield */
  value?: any;
  /**Id for testing */
  datatestid?: string;
}

/**TextField of the todo-app application */
export const TextField: React.FC<TextFieldProps> = ({
  labelText,
  ...props
}): JSX.Element => {
  const [messageError, setMessageError] = useState<string>("");

  const handleKeyPress = (event: KeyboardEvent<HTMLInputElement>) => {
    if (props.type === "numbers" && !/[0-9]/.test(event.key)) {
      setMessageError("Text is not allowed");
      event.preventDefault();
    } else {
      setMessageError("");
    }
  };

  return (
    <div
      data-testid={props.datatestid}
      className={`${props.variant}-container`}
    >
      <label htmlFor={props.id}>
        <p className={`${props.variant}-label`}>
          <b>{labelText}</b>
        </p>
        <input
          {...props}
          onKeyPress={handleKeyPress}
          className={`textField ${props.className}`}
          name={props.name}
        />
      </label>
      <div className="error">
        <MessageError size="small">{messageError}</MessageError>
      </div>
    </div>
  );
};

export default TextField;
