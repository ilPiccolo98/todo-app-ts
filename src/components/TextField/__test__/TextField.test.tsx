import React, { ChangeEvent } from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import TextField from "../TextField";

const renderTextField = (
  type: "text" | "numbers" | "password",
  onChange: (e: ChangeEvent<HTMLInputElement>) => void
) => {
  return render(
    <TextField
      datatestid="textfield"
      variant="primary"
      labelText="Testing"
      type={type}
      onChange={onChange}
    />
  );
};

const getLabelTextBox = (textbox: HTMLElement): HTMLLabelElement => {
  return textbox.childNodes[0] as HTMLLabelElement;
};

const getInputTextBox = (textbox: HTMLElement): HTMLInputElement => {
  const label: HTMLLabelElement = getLabelTextBox(textbox);
  return label.childNodes[1] as HTMLInputElement;
};

const getMessageError = (textbox: HTMLElement): HTMLParagraphElement => {
  return textbox.childNodes[
    textbox.childElementCount - 1
  ] as HTMLParagraphElement;
};

const handleChangeObject = {
  handleChange: (e: ChangeEvent<HTMLInputElement>) => {},
};

const handleChangeSpy = jest.spyOn(handleChangeObject, "handleChange");

describe("Testing Textfield component", () => {
  beforeEach(() => {
    handleChangeSpy.mockReset();
  });

  it("should call the handleChange event and the value should be 'test' by using type=text", () => {
    const component = renderTextField("text", handleChangeObject.handleChange);
    const textbox = component.getByTestId("textfield");
    const input = getInputTextBox(textbox);
    fireEvent.change(input, {
      target: {
        value: "test",
      },
    });
    expect(handleChangeSpy).toHaveBeenCalledTimes(1);
    expect(input.value).toBe("test");
    expect(component).toMatchSnapshot();
  });

  it("should call the handleChange event and the value should be '123' by using type=numbers", () => {
    const component = renderTextField(
      "numbers",
      handleChangeObject.handleChange
    );
    const textbox = component.getByTestId("textfield");
    const input = getInputTextBox(textbox);
    fireEvent.change(input, {
      target: {
        value: "123",
      },
    });
    expect(handleChangeSpy).toHaveBeenCalledTimes(1);
    expect(input.value).toBe("123");
    expect(component).toMatchSnapshot();
  });

  it("should call the handleChange event and the value should be 'pass' by using type=password", () => {
    const component = renderTextField(
      "password",
      handleChangeObject.handleChange
    );
    const textbox = component.getByTestId("textfield");
    const input = getInputTextBox(textbox);
    fireEvent.change(input, {
      target: {
        value: "pass",
      },
    });
    expect(handleChangeSpy).toHaveBeenCalledTimes(1);
    expect(input.value).toBe("pass");
    expect(component).toMatchSnapshot();
  });

  it("should show the message 'Text is not allowed' if I write text when the type=numbers", () => {
    const component = renderTextField(
      "numbers",
      handleChangeObject.handleChange
    );
    const textbox = component.getByTestId("textfield");
    const input = getInputTextBox(textbox);
    fireEvent.keyPress(input, {
      key: "a",
      code: "KeyA",
      charCode: 65,
    });
    const messageError = getMessageError(textbox);
    expect(messageError.textContent).toBe("Text is not allowed");
    expect(component).toMatchSnapshot();
  });

  it("shouldn't show the message 'Text is not allowed' if I write text when the type=text", () => {
    const component = renderTextField("text", handleChangeObject.handleChange);
    const textbox = component.getByTestId("textfield");
    const input = getInputTextBox(textbox);
    fireEvent.keyPress(input, {
      key: "a",
      code: "KeyA",
      charCode: 65,
    });
    const messageError = getMessageError(textbox);
    expect(messageError.textContent).not.toBe("Text is not allowed");
    expect(component).toMatchSnapshot();
  });
});
