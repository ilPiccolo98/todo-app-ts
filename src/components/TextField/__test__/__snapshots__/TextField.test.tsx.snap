// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Textfield component should call the handleChange event and the value should be '123' by using type=numbers 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="primary-container"
        data-testid="textfield"
      >
        <label>
          <p
            class="primary-label"
          >
            <b>
              Testing
            </b>
          </p>
          <input
            class="textField undefined"
            datatestid="textfield"
            type="numbers"
            variant="primary"
          />
        </label>
        <div
          class="error"
        >
          <span
            class="message-error undefined small"
          >
            <b />
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="primary-container"
      data-testid="textfield"
    >
      <label>
        <p
          class="primary-label"
        >
          <b>
            Testing
          </b>
        </p>
        <input
          class="textField undefined"
          datatestid="textfield"
          type="numbers"
          variant="primary"
        />
      </label>
      <div
        class="error"
      >
        <span
          class="message-error undefined small"
        >
          <b />
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Testing Textfield component should call the handleChange event and the value should be 'pass' by using type=password 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="primary-container"
        data-testid="textfield"
      >
        <label>
          <p
            class="primary-label"
          >
            <b>
              Testing
            </b>
          </p>
          <input
            class="textField undefined"
            datatestid="textfield"
            type="password"
            variant="primary"
          />
        </label>
        <div
          class="error"
        >
          <span
            class="message-error undefined small"
          >
            <b />
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="primary-container"
      data-testid="textfield"
    >
      <label>
        <p
          class="primary-label"
        >
          <b>
            Testing
          </b>
        </p>
        <input
          class="textField undefined"
          datatestid="textfield"
          type="password"
          variant="primary"
        />
      </label>
      <div
        class="error"
      >
        <span
          class="message-error undefined small"
        >
          <b />
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Testing Textfield component should call the handleChange event and the value should be 'test' by using type=text 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="primary-container"
        data-testid="textfield"
      >
        <label>
          <p
            class="primary-label"
          >
            <b>
              Testing
            </b>
          </p>
          <input
            class="textField undefined"
            datatestid="textfield"
            type="text"
            variant="primary"
          />
        </label>
        <div
          class="error"
        >
          <span
            class="message-error undefined small"
          >
            <b />
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="primary-container"
      data-testid="textfield"
    >
      <label>
        <p
          class="primary-label"
        >
          <b>
            Testing
          </b>
        </p>
        <input
          class="textField undefined"
          datatestid="textfield"
          type="text"
          variant="primary"
        />
      </label>
      <div
        class="error"
      >
        <span
          class="message-error undefined small"
        >
          <b />
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Testing Textfield component should show the message 'Text is not allowed' if I write text when the type=numbers 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="primary-container"
        data-testid="textfield"
      >
        <label>
          <p
            class="primary-label"
          >
            <b>
              Testing
            </b>
          </p>
          <input
            class="textField undefined"
            datatestid="textfield"
            type="numbers"
            variant="primary"
          />
        </label>
        <div
          class="error"
        >
          <span
            class="message-error undefined small"
          >
            <b>
              Text is not allowed
            </b>
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="primary-container"
      data-testid="textfield"
    >
      <label>
        <p
          class="primary-label"
        >
          <b>
            Testing
          </b>
        </p>
        <input
          class="textField undefined"
          datatestid="textfield"
          type="numbers"
          variant="primary"
        />
      </label>
      <div
        class="error"
      >
        <span
          class="message-error undefined small"
        >
          <b>
            Text is not allowed
          </b>
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Testing Textfield component shouldn't show the message 'Text is not allowed' if I write text when the type=text 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="primary-container"
        data-testid="textfield"
      >
        <label>
          <p
            class="primary-label"
          >
            <b>
              Testing
            </b>
          </p>
          <input
            class="textField undefined"
            datatestid="textfield"
            type="text"
            variant="primary"
          />
        </label>
        <div
          class="error"
        >
          <span
            class="message-error undefined small"
          >
            <b />
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="primary-container"
      data-testid="textfield"
    >
      <label>
        <p
          class="primary-label"
        >
          <b>
            Testing
          </b>
        </p>
        <input
          class="textField undefined"
          datatestid="textfield"
          type="text"
          variant="primary"
        />
      </label>
      <div
        class="error"
      >
        <span
          class="message-error undefined small"
        >
          <b />
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
